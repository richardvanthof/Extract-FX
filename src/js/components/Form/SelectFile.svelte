<script lang="ts">
    // Importing the store
    import { sourceData,  } from '../../global-vars/ingest';

    // Define the type for the source data
    type SourceData = {
        type: 'RS-FX-EXCHANGE';
        track: number;
        sequence: string;
        exclusions: string[];
        clips: object[];
    };

    let files:File[]|null|undefined = $state();

    // Initialize state for file content
    let fileContent: SourceData | null = null;

    // Handle file input change
    const handleFile = (event: Event) => {

        const fileInput = event.target as HTMLInputElement;
        const file:File|null|undefined = files && files[0]; // Get the first file selected

        if (file) {
            // Create a FileReader to read the file
            const reader = new FileReader();

            // FileReader event to handle the file once it's read
            reader.onload = () => {
                const content: string = reader.result as string;

                try {
                    // Try to parse JSON content
                    const parsedContent: SourceData = JSON.parse(content);

                    // Check if the file is empty or does not contain valid data
                    if (content.length === 0) {
                        throw new Error('This file is empty!');
                    }

                    if (parsedContent.type !== 'RS-FX-EXCHANGE' || parsedContent.clips.length <= 0) {
                        throw new Error('This file is invalid and cannot be read by this program.');
                    }

                    // Update the store with the file content
                    sourceData.set(parsedContent);
                    
                } catch (err: any) {
                    // If JSON parsing fails, show an error message
                    fileContent = null;
                    sourceData.set(null)
                    files = [];
                    alert(err.message);
                }
            };

            // Read the file as text
            reader.readAsText(file);
        } else {
            // Clear content if no file is selected
            fileContent = null;
        }
    };


</script>

<style>
    input {
        background: var(--bg-dark);
        border: var(--border-styling);
        border-radius: 3px;
        color: white;
        padding: .3em;
        display: block;
        width: 100%;
        margin-bottom: 1em;
        cursor: pointer;
    }

    pre {
        background: #f4f4f4;
        padding: 1em;
        border-radius: 4px;
        white-space: pre-wrap;
        word-wrap: break-word;
    }
</style>

<!-- File input field that accepts only .json files -->
<input bind:files onchange={handleFile} accept="application/json" type="file">
