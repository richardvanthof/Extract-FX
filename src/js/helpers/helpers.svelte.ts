const handleClick = (e:Event, callback: unknown) => {
    e.preventDefault()
    return callback;
}

const generateNumberedOptions = (amount: number, prefix?: string, suffix?: string):[string,number][] => {
    const options:[string,number][] = [];
    for (let index = 1; index <= amount; index++) {
        options.push([
            `${prefix ? prefix : ''} ${index} ${suffix ? suffix : ''}`,
            index
        ]);
    }

    return options;
}
const getUniqueKeys = (data:object[]|object):string[] => {
    let keys:string[];
    if(Array.isArray(data)) { keys = data.map((val) => Object.keys(val)).flat();} 
    else { keys = Object.keys(data);};

    return keys.filter((value, index, array) => {
        const normalizedArray = array.map((val) => val.toLocaleLowerCase()) 
        return normalizedArray.indexOf(value.toLocaleLowerCase()) === index;
    }); 
};

export {handleClick, getUniqueKeys, generateNumberedOptions}