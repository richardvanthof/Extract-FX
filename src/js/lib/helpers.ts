const handleClick = (e:Event|null|undefined, callback: any) => {
    e && e.preventDefault()
    return callback;
}

export type NumberedOption  = [string|number, number];

const generateNumberedOptions = (amount: number, prefix?: string, suffix?: string):NumberedOption[] => {
    let options:NumberedOption[] = [];
    for (let index = 1; index <= amount; index++) {
        options.push([
            `${prefix ? prefix : ''} ${index} ${suffix ? suffix : ''}`,
            index
        ]);
    }

    return options;
}

export {handleClick, generateNumberedOptions}