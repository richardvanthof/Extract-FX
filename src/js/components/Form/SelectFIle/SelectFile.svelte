<script lang="ts">
    // Importing the store
    import { globals } from '@/js/global-vars/globals.svelte';  // Assuming this store is used to store data globally
    import { getUniqueKeys, getJSON } from './SelectFile.helpers';  // Utility for extracting keys from the file if needed
    import type {SourceData} from './SelectFile.helpers';
    import type {Writable} from 'svelte/store';
    import {get} from 'svelte/store';
    import {v4 as uuid} from 'uuid';

    // Declare a variable to hold the files
    let files: FileList | null = null;

    // Handle file input change
    const handleFile = async (event: Event) => {
        const input = event.target as HTMLInputElement;
        const files = input?.files;

        try {
            // Decode and parse the JSON file
            if(!files) {throw new Error("File not found.")}
            const decodedData: SourceData = await getJSON(files[0]);
        
            // Set the decoded data in the store
            sourceData.set(decodedData);
            
            const uniqueKeys = getUniqueKeys(decodedData.clips);
            
            // Optionally: If you want to extract unique keys from the data, you can use the helper function
            if(uniqueKeys) {exclusionOptions.set(uniqueKeys);}

            // Add all the exclusions from the source file to the exclude panel.
            if(decodedData && exclusions) {
                exclusions.set(decodedData.exclusions.map((exclusion) => ({
                    effect: exclusion,
                    id: uuid()
                })))
            };

        } catch (err) {
            console.error(err);
            alert(err)
        }
    };
</script>


<style>
    input {
        background: var(--bg-dark);
        border: var(--border-styling);
        border-radius: var(--border-radius-inner);
        color: var(--white);
        padding: .3em;
        display: block;
        width: 100%;
        margin-bottom: 1em;
        cursor: pointer;
    }
</style>

<!-- File input field that accepts only .json files -->
<input bind:files={files} onchange={handleFile} accept="application/json" type="file">
