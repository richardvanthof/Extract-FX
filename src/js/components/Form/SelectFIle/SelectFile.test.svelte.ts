import { describe, expect, it, vi } from 'vitest';
import {screen, render} from '@testing-library/svelte';
import { userEvent } from '@testing-library/user-event';
import { getJSON, createExclusions, handleIngestFile } from './SelectFile.helpers.svelte';
import type {FileData} from './SelectFile.helpers.svelte';
import SelectFile from './SelectFile.svelte';
import type { SourceData } from '@/js/global-vars/globals.svelte';
import '@testing-library/jest-dom';

describe("getJSON", async () => {
    it("Returns JS object from JSON-file", async () => {
        const data:SourceData = {
            type: 'RS-FX-EXCHANGE',
            track: 1,
            sequence: 'My Video',
            exclusions: [],
            clips: [
                {Motion: {}, Opacity: {}},
                {Motion: {}, Opacity: {}},
                {Motion: {}, Opacity: {}}
            ]
        };

        const file = new File([JSON.stringify(data)], 'effectsList.json', {
            type: 'application/json',
        });

        const result = await getJSON(file);
        
        expect(result).toStrictEqual(data);
    });

    it("Returns error when file is empty.", async () => {
        const file = new File([''], 'emptyList.json', {
            type: 'application/json',
        });

        await expect(getJSON(file)).rejects.toThrowError("This file is empty.");
    })

    it("Returns error when file does not contain clips data.", async () => {
        const data:SourceData= {
            type: 'RS-FX-EXCHANGE',
            track: 1,
            sequence: 'My Video',
            exclusions: [],
            clips: []
        };

        const file = new File([JSON.stringify(data)], 'noClips.json', {
            type: 'application/json',
        });

        await expect(getJSON(file)).rejects.toThrowError("No clip effect data has been found.");
    })

    it("Returns error when file does not contain the correct data type.", async () => {
        const data:SourceData= {
            // @ts-expect-error part of test
            type: 'NOT-RS-FX-EXCHANGE', // this file is not the correct type.
            track: 1,
            sequence: 'My Video',
            exclusions: [],
            clips: [
                {Motion: {}, Opacity: {}},
                {Motion: {}, Opacity: {}},
                {Motion: {}, Opacity: {}}
            ]
        };

        const file = new File([JSON.stringify(data)], 'invalid.json', {
            type: 'application/json',
        });

        await expect(getJSON(file)).rejects.toThrowError("This file contains invalid data and cannot be opened.");
    })
})

describe("handleIngestFile", async () => {
    it("Returns JS object from JSON-file", async () => {
        const data:SourceData = {
            type: 'RS-FX-EXCHANGE',
            track: 1,
            sequence: 'My Video',
            exclusions: [],
            clips: [
                {Motion: {}, Opacity: {}},
                {Motion: {}, Opacity: {}},
                {Motion: {}, Opacity: {}}
            ]
        };

        const resultObj:FileData = {
            data,
            exclusionOptions: ['Motion', 'Opacity']
        }
        
        const file = new File([JSON.stringify(data)], 'effectsList.json', {
            type: 'application/json',
        });

        let filesList:FileData|null = $state(null);
        
        const props = {
            error: null,
            label: 'source',
            callback: vi.fn().mockImplementation(async (files) => {
                filesList = files
            })
        }
        
        const user = userEvent.setup()

        const {getByLabelText} = render(SelectFile, {props})
        const input = getByLabelText(props.label);

        await user.upload(input, file)
        if(filesList != null) {
            expect(await handleIngestFile(filesList)).toStrictEqual(resultObj)
        }
    });

    it("Puts error under file input field", async () => {
        const data:SourceData = {
            type: 'RS-FX-EXCHANGE',
            track: 1,
            sequence: 'My Video',
            exclusions: [],
            clips: [
                {Motion: {}, Opacity: {}},
                {Motion: {}, Opacity: {}},
                {Motion: {}, Opacity: {}}
            ]
        };
        
        const file = new File([JSON.stringify(data)], 'effectsList.json', {
            type: 'application/json',
        });
        
        const props = {
            error: null,
            label: 'source',
            callback: vi.fn()
        }
        
        const user = userEvent.setup()

        const {getByLabelText, rerender} = render(SelectFile, {props})
        const input = getByLabelText(props.label);


        await user.upload(input, file)
        let error:TypeError|null = null;
        
        // @ts-expect-error: parameter should be undefined for test.
        await handleIngestFile(null).catch(err => error = err)
        
        await rerender({error});

        const errorField = screen.getByTestId('select-file-error')
        
        // Check if error is displayed under file input field
        if(error != null) {
            // @ts-expect-error part of test
            expect(error.message).toBe("Cannot read properties of null (reading 'target')");
            expect(errorField).toBeInTheDocument(); 

        }
    });

})

describe("Create exclusions", () => {
    it("Transform exclusions to format compactible with Dropdown component.", ()=> {
        const input = ["Warp Stabilizer", "Motion", "Displace"];

        const result = createExclusions(input);
        
        // Check if the result is an array with the same length as input
        expect(result).toHaveLength(input.length);

        // Check each object in the array
        result.forEach((item, index) => {
            expect(item).toHaveProperty('effect', input[index]);
            expect(item).toHaveProperty('id');
            expect(item.id).toMatch(
            /^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i
            );  // Validate UUID format
        });
    })
})

