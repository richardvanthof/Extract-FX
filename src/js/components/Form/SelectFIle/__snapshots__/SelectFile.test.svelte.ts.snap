// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`handleIngestFile > Puts error under file input field 1`] = `
{
  "$destroy": [Function],
  "$on": [Function],
  "$set": [Function],
}
`;
